### PR TITLE
fix: `tinacms/dist/cache` not found error

### PR DESCRIPTION
**Pull request by kldavis4**
Fix issue #4773 

Switch to dynamic imports to load the cache module so that webpack loads it properly. Also fix the cache directory to use /tmp when needed.

### CODE REVIEW

File: .changeset/rare-turkeys-bow.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'tinacms': patch
+---
+
+Fix module not found error at runtime for tinacms/dist/cache
```
Comment (kldavis4) [inline .changeset/rare-turkeys-bow.md:5]:
```suggestion
Fix module not found `tinacms/dist/cache` error at runtime due to broken webpack bundling
```

File: packages/tinacms/src/cache/node-cache.ts [hunk 2]
```diff
@@ -6,12 +6,31 @@ const makeKey = (key: any) => {
   return createHash('sha256').update(input).digest('hex')
 }
 
-export const NodeCache = (dir: string, fs: any): Cache => {
+// makeCacheDir creates the cache directory if it doesn't exist
+const makeCacheDir = (dir: string, fs: any) => {
+  const path = require('path')
+  const os = require('os')
```
Comment (JackDevAU) [inline packages/tinacms/src/cache/node-cache.ts:12]: Why do we need to use `require` here?

File: packages/tinacms/src/cache/node-cache.ts [hunk 3]
```diff
@@ -6,12 +6,31 @@ const makeKey = (key: any) => {
...
+  if (!fs.existsSync(parts[0])) {
+    cacheDir = path.join(os.tmpdir(), parts[0])
+  }
+
+  fs.mkdirSync(cacheDir, { recursive: true })
+  return cacheDir
+}
+
+export const NodeCache = (dir: string): Cache => {
+  const fs = require('fs')
```
Comment (JackDevAU) [inline packages/tinacms/src/cache/node-cache.ts:27]: Why do we need to use `require` here?

File: packages/tinacms/src/unifiedClient/index.ts [hunk 4]
```diff
@@ -47,18 +50,29 @@ export class TinaClient<GenQueries> {
...
-    // if we're in node, use the fs cache
-    if (cacheDir && typeof require !== 'undefined') {
-      const { NodeCache } = require('tinacms/dist/cache')
-      this.cache = NodeCache(cacheDir, require('fs'))
+  async init() {
+    if (this.initialized) {
+      return
+    }
+    try {
+      if (this.cacheDir && typeof require !== 'undefined') {
```
Comment (JackDevAU) [inline packages/tinacms/src/unifiedClient/index.ts:61]: Why do we need to check if `require` is undefined here?

Comment (kldavis4) [reply packages/tinacms/src/unifiedClient/index.ts:61]: Detects if nodejs environment which has fs/path/os apis (versus edge and browser runtime)

File: packages/tinacms/src/cache/node-cache.ts (hunk 2 above)
Comment (kldavis4) [reply packages/tinacms/src/cache/node-cache.ts:12]: This is conditionally imported depending on whether it is a nodejs env.

The only alternative (without polyfilling) is dynamic imports - I wasn't able to get them working correctly though

**Comment by kldavis4** [comment]
pushed some changes to allow dynamic imports to be used. There's still an issue with the cli dev/build commands on this project: https://github.com/comboomPunkTsucht/comboom.sucht-BLOG

Not sure what exactly is the cause yet though

**Comment by kldavis4** [comment]
This also includes CLI updates to fix #4788

File: packages/tinacms/src/cache/node-cache.ts [hunk 5]
```diff
@@ -1,17 +1,36 @@
...
-  return createHash('sha256').update(input).digest('hex')
+// makeCacheDir creates the cache directory if it doesn't exist
+const makeCacheDir = async (dir: string, fs: any) => {
+  const path = await import('path')
+  const os = await import('os')
+
+  const parts = dir.split(path.sep).filter(Boolean)
+
+  let cacheDir = dir
+  // check if the root directory exists, if not then create create in tmp and return new path
```
Comment (ncn-ssw) [inline packages/tinacms/src/cache/node-cache.ts:11]: **nitpick**: Should the second 'create' be 'cache'?
```suggestion
  // check if the root directory exists, if not then create cache in tmp and return new path
```

**Review by ncn-ssw — APPROVED** [review]
🦙 Approved

**Review by JackDevAU — APPROVED** [review]
🦙

**Pull request merged**